exec/java-exec/src/main/java/org/apache/drill/exec/metastore/analyze/AnalyzeInfoProviderRegistry.java

```diff
@@ -0,0 +1,128 @@
+
+  public AnalyzeInfoProviderRegistry(ScanResult classpathScan) {
+    this.classpathScan = classpathScan;
+  }
+
+  /**
+   * Returns {@link AnalyzeInfoProvider} instance suitable for specified {@link GroupScan}.
+   *
+   * @param groupScan group scan
+   * @return {@link AnalyzeInfoProvider} instance
+   */
+  public AnalyzeInfoProvider get(GroupScan groupScan) {
+    if (analyzeInfoProviders == null) {
+      synchronized (this) {
+        if (analyzeInfoProviders == null) {
+          analyzeInfoProviders = Collections.unmodifiableList(init());
+        }
+      }
+    }
+
+    List<AnalyzeInfoProvider> suitableProviders = analyzeInfoProviders.stream()
+        .filter(analyzeInfoProvider -> analyzeInfoProvider.supportsGroupScan(groupScan))
+        .collect(Collectors.toList());
+
+    if (suitableProviders.isEmpty()) {
+      throw new MetastoreException(String.format("Unable to find AnalyzeInfoProvider implementation for GroupScan class [%s]",
+          groupScan.getClass().getSimpleName()));
+    } else if (suitableProviders.size() > 1) {
+      logger.warn("Found several AnalyzeInfoProvider implementations for the same GroupScan class [{}}]. " +
+          "Please specify only unique AnalyzeInfoProvider implementation for every group scan using [{}] annotation " +
+              "to prevent using arbitrary implementation.",
+          groupScan.getClass().getSimpleName(), AnalyzeInfoProviderTemplate.class.getName());
+    }
+
+    return suitableProviders.get(0);
+  }
+
+  private List<AnalyzeInfoProvider> init() {
+    List<AnalyzeInfoProvider> providers;
+    List<AnnotatedClassDescriptor> annotatedClasses = classpathScan.getAnnotatedClasses(AnalyzeInfoProviderTemplate.class.getName());
+
+    providers = annotatedClasses.stream()
+        .map(AnnotatedClassDescriptor::getClassName)
+        .map(this::getAnalyzeInfoProviderInstance)
+        .filter(Objects::nonNull)
+        .map(instance -> (AnalyzeInfoProvider) instance)
+        .collect(Collectors.toList());
+    return providers;
+  }
+
+  private Object getAnalyzeInfoProviderInstance(String className) {
+    MethodHandles.Lookup publicLookup = MethodHandles.publicLookup();
+    MethodHandle constructor;
+    try {
+      MethodType methodType = MethodType.methodType(void.class);
+      constructor = publicLookup.findConstructor(Class.forName(className), methodType);
+    } catch (ClassNotFoundException e) {
+      logger.error("Unable to find AnalyzeInfoProvider implementation class [{}}]", className, e);
+      return null;
+    } catch (NoSuchMethodException | IllegalAccessException e) {
+      logger.error("AnalyzeInfoProvider implementation class [%s] must have constructor without arguments.");
+      return null;
+    }
+
+    Object instance;
+    try {
+      instance = constructor.invoke();
+    } catch (Throwable e) {
+      logger.error("Unable to init AnalyzeInfoProvider class [{}}]", className, e);
```

Comment:
I think warn is more suitable and does not require stackstace. Can only have error message from throwable. Full stackstace can be added in debug level.